store/postgres/src/chain_store.rs

```diff
@@ -1813,6 +1841,120 @@ impl ChainStoreTrait for ChainStore {
     }
 }
 
+mod recent_blocks_cache {
+    use super::*;
+
+    type OffsetFromChainHead = BlockNumber;
+
+    struct Inner {
+        blocks: HashMap<OffsetFromChainHead, (BlockPtr, Option<json::Value>)>,
+        capacity: usize,
+    }
+
+    /// We cache the most recent blocks in memory to avoid overloading the
+    /// database with unnecessary queries close to the chain head. We invalidate
+    /// blocks:
+    /// - After 5 seconds.
+    /// - Whenever the chain head advances.
+    pub struct RecentBlocksCache {
+        // We protect everything with a global `RwLock` to avoid data races. Ugly...
+        inner: RwLock<Inner>,
+    }
+
+    impl RecentBlocksCache {
+        pub fn new(capacity: usize) -> Self {
+            RecentBlocksCache {
+                inner: RwLock::new(Inner {
+                    blocks: HashMap::new(),
+                    capacity,
+                }),
+            }
+        }
+
+        pub fn set_chain_head(&self, chain_head: BlockPtr) {
+            // We invalidate the entire cache when the chain head changes.
+            let mut inner = self.inner.write();
+            inner.blocks.clear();
+            inner.blocks.insert(0, (chain_head, None));
+        }
+
+        pub fn chain_head_ptr(&self) -> Option<BlockPtr> {
+            self.inner.read().blocks.get(&0).map(|p| p.0.clone())
+        }
+
+        pub fn clear(&self) {
+            self.inner.write().blocks.clear();
+        }
+
+        fn chain_head_number_inner(inner: &Inner) -> Option<BlockNumber> {
+            inner.blocks.get(&0).map(|p| p.0.number)
+        }
+
+        pub fn get_block(
+            &self,
+            ancestor: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(BlockPtr, Option<json::Value>)> {
+            Self::get_block_inner(&*self.inner.read(), ancestor, offset)
+        }
+
+        fn get_block_inner(
+            inner: &Inner,
+            ancestor: &BlockPtr,
+            offset: BlockNumber,
+        ) -> Option<(BlockPtr, Option<json::Value>)> {
```

Comment:
This is used in a few places to check if that block is in the cache; it has to clone the cached value because of this return type. It would be better to either return references here or to have a separate `has_block_inner` method for places where we just need to check if a block is cached.

Reply:
I added a `contains_blocks` method on `Inner` (so it's not available in the public API, but it can be made public if we need it in other places). Unfortunately we can't easily return references (we could do `Arc`'s though) because of locking.

Reply:
I think using an `Arc<Option<json::Value>>` would be a good thing, but it will require lots of changes across the code base. It would be a very good as a separate follow-on to this PR